federatedscope/autotune/pfedhpo/fl_server.py

```diff
@@ -0,0 +1,374 @@
+        if not self.discrete:
+            for k, v in self._ss.items():
+                if not (hasattr(v, 'lower') and hasattr(v, 'upper')):
+                    raise ValueError("Unsupported hyper type {}".format(type(v)))
+                else:
+                    if v.log:
+                        l, u = np.log10(v.lower), np.log10(v.upper)
+                    else:
+                        l, u = v.lower, v.upper
+                    self.pbounds[k] = (l, u)
+        else:
+            for k, v in self._ss.items():
+                if not (hasattr(v, 'lower') and hasattr(v, 'upper')):
+                    if hasattr(v, 'choices'):
+                        self.pbounds[k] = list(v.choices)
+                    else:
+                        raise ValueError("Unsupported hyper type {}".format(type(v)))
+                else:
+                    if v.log:
+                        l, u = np.log10(v.lower), np.log10(v.upper)
+                    else:
+                        l, u = v.lower, v.upper
+                    N_samp = 10
+                    samp = []
+                    for i in range(N_samp):
+                        samp.append((u - l) / N_samp * i + l)
+                    self.pbounds[k] = samp
+
+        # prepare hyper-net
+        self.client2idx = None
+
+        if not self.train_anchor:
+            hyper_enc = torch.load(os.path.join(self._cfg.hpo.working_folder,
+                                                'hyperNet_encoding.pt'))
+            if self._cfg.data.type == 'mini-graph-dc':
+                dim = 60
+            elif 'cifar' in str(self._cfg.data.type).lower():
+                dim = 60
+            elif 'femnist' in str(self._cfg.data.type).lower():
+                dim = 124
+            else:
+                raise NotImplementedError
+
+            self.client_encoding = torch.ones(client_num, dim)
+            if not self.discrete:
+                self.HyperNet = HyperNet(encoding=self.client_encoding,
+                                         num_params=len(self.pbounds),
+                                         n_clients=client_num,
+                                         device=self._cfg.device,
+                                         var=0.01).to(self._cfg.device)
+            else:
+                self.HyperNet = DisHyperNet(encoding=self.client_encoding,
+                                            cands=self.pbounds,
+                                            n_clients=client_num,
+                                            device=self._cfg.device,
+                                            ).to(self._cfg.device)
+
+            self.HyperNet.load_state_dict(hyper_enc['hyperNet'])
+            self.HyperNet.eval()
+            if not self.discrete:
+                self.raw_params = self.HyperNet()[0].detach().cpu().numpy()
+            else:
+                self.logits = self.HyperNet()[0]
+
+    def callback_funcs_model_para(self, message: Message):
```

Comment:
is it necessary to have this branch: https://github.com/alibaba/FederatedScope/blob/b3f823d138e574b070704017b934fb1fdb2750d5/federatedscope/core/workers/server.py#L900

Reply:
not necessary